core/include/irq.h

```diff
@@ -60,6 +61,14 @@ unsigned irq_enable(void);
  */
 void irq_restore(unsigned state);
 
+/**
+ * @brief   Test if IRQs are currently enabled
+ *
+ * @return  0 (false) if IRQs are currently disabled
+ * @return  != 0 (true) if IRQs are currently enabled
```

Comment:
I think the API should clearly state what to expect if this function is called from interrupt context.

Reply:
Does it make a difference?

Reply:
Some would argue in ISR no other interrupts will interrupt the ISR and always return `0`, even when interrupts are not blocked by anything else then the fact an ISR is currently handeled. Others would disagree on that. When people would rely on either behaviour, bad things happen when the actual behaviour doesn't match the expectation. Also, maybe on some hardware it might be impossible to get the result from ISR context.

Stating explicitly that the return value is undefined when called from interrupt context might be the safest thing to do here.

Reply:
I'm not sure if an undefined return value would really be the best choice? Either the architecture implicitly disables the interrupts when an ISR is entered, or the architecture also allows checking the state within an ISR.

Reply:
The least overhead would be to behave the same regardless of context for all platforms this function would just access a special register, provided that register will still return meaningful data from interrupt context. (Otherwise it would still be efficient, but not useful.)

Reply:
I just learned that Interrupts on ARM can be interrupted by another interrupt if configured that way. I was aware that ARM has a way cooler interrupt implementation that e.g. AVR and MSP430. But I'm surprised by how far more sophisticated their stuff is.

Keeping this in mind this really seems to be a pitfall. E.g. lets say my ISR wants to check wether it could be interrupted by another interrupt. I think this function should return false if not, and yes if it could be interrupted. Following that logic, on AVR this function should always return false in ISR context. Right?

Reply:
> Suffice it to say that if you’ve got two USARTs on a system and you want either USART to interrupt the SPI ISR but not each other, and you’d like USART0 to always get handled before USART1 for instance, you can do it.

Original source: https://hackaday.com/2015/09/18/embed-with-elliot-interrupts-the-good/

Reply:
BTW, xtensa processors such as the ESP8266 and the ESP32 also allow the interruption of an ISR if interrupts are not explicitly disabled. Disabling interrupts is done explicitly in software in ISR hooks.

Reply:
It seems I was wrong again. On AVR interrupts *can* interrupt an ISR. But upon serving an interrupt the AVR unsets the interrupt enable bit. Reenabling interrupts during ISR would allow getting interrupted. The problem for AVRs is that they are particularly short on ROM and slow, so that the stack on which the state to resume on after the ISR is stored easily overflows. Thus, it was decided to disable interrupts during ISR by default. This also means that on the AVR the trivial implementation would behave as expected.

Reply:
I'd like again point to the hackaday blog, the second article in the series gives a code example on how to declare an ISR on AVR that can be interrupted:

https://hackaday.com/2015/09/25/embed-with-elliot-interrupts-the-bad/

Reply:
RIOT needs a lot of architecture dependent care, should nested interrupts be enabled. We decided at some point that it's just not worth the effort to support that, and that we'd just not put unnecessary barriers in place, so application developers can hack in nested interrupts if they know what they're doing.

Reply:
what he says :-)

Reply:
After reading this again, the documentation is actually quite specific. You might want to use `@retval` over `@return`. (Be careful to not add a space between (`=!` and `0`, if you use `@retval`, as it Doxygen will use the "first word" as return value.)